contracts/infrastructure/dapp/paraswap/ParaswapFilter.sol

```diff
@@ -269,14 +280,49 @@ contract ParaswapFilter is BaseFilter {
         uint256 exchangeDataOffset = 36 + abi.decode(_data[196:228], (uint256)); 
         address[] memory spenders = new address[](_callees.length);
         bytes[] memory allData = new bytes[](_callees.length);
-        for(uint256 i = 0; i < _callees.length; i++) {
+        uint256 i;
+        while(i < _callees.length) {
```

Comment:
I think this doesn't work as you are looping over `callees.length` but you are reducing the length if the callee is augustus so you might never reach the last callee. Also you are not incrementing `i` when you reduce the size of `callees` so you will check the same callee twice. Maybe I'm wrong but it's just a sign that this logic is too complicated.

A simpler version of the same logic that I find simpler to read is
```
function hasAuthorisedCallees(
        address _augustus,
        address[] memory _callees,
        uint256[] memory _startIndexes,
        bytes calldata _data
    )
        internal
        view
        returns (bool)
    {
        // _data = {sig:4}{six params:192}{exchangeDataOffset:32}{...}
        // we add 4+32=36 to the offset to skip the method sig and the size of the exchangeData array
        uint256 exchangeDataOffset = 36 + abi.decode(_data[196:228], (uint256)); 
        address[] memory to = new address[](_callees.length);
        address[] memory spenders = new address[](_callees.length);
        bytes[] memory allData = new bytes[](_callees.length);
        uint256 i;
        while(i < _callees.length) {
            bytes calldata slicedExchangeData = _data[exchangeDataOffset+_startIndexes[i] : exchangeDataOffset+_startIndexes[i+1]];
            if (_callees[i] == _augustus) {
                if (slicedExchangeData.length >= 4 && getMethod(slicedExchangeData) == WITHDRAW_ALL_WETH) {
                    uint newLength = _callees.length - 1;
                    // solhint-disable-next-line no-inline-assembly
                    assembly {
                        mstore(to, newLength)
                        mstore(spenders, newLength)
                        mstore(allData, newLength)
                    }
                } else {
                    return false;
                }
            } else {
                to[i] = _callees[i];
                allData[i] = slicedExchangeData;
                spenders[i] = Utils.recoverSpender(_callees[i], slicedExchangeData);
            }
            i++;
        }
        return authoriser.areAuthorised(_augustus, spenders, to, allData);
    }
 ```

Reply:
Agreed that using a temporary array for callees is much cleaner